File: experiments/adrianco/experiment-5/runs/language=typescript_model=claude-opus-4-7_tooling=beads/rep2/src/queries/players.ts
import type { Player } from "../types.js";
import { normalizeTeam, stripDiacritics } from "../normalize.js";

export interface PlayerFilter {
  name?: string;
  nationality?: string;
  club?: string;
  position?: string;
  minOverall?: number;
  maxOverall?: number;
  minAge?: number;
  maxAge?: number;
  sortBy?: "overall" | "potential" | "age" | "name";
  sortDir?: "asc" | "desc";
  limit?: number;
}

function ci(haystack: string, needle: string): boolean {
  return stripDiacritics(haystack).toLowerCase().includes(stripDiacritics(needle).toLowerCase());
}

export function findPlayers(players: Player[], filter: PlayerFilter): Player[] {
  const clubN = filter.club ? normalizeTeam(filter.club) : undefined;

  let res = players.filter((p) => {
    if (filter.name && !ci(p.name, filter.name)) return false;
    if (filter.nationality && !ci(p.nationality, filter.nationality)) return false;
    if (clubN && p.clubNormalized !== clubN && !ci(p.club, filter.club!)) return false;
    if (filter.position && !ci(p.position ?? "", filter.position)) return false;
    if (filter.minOverall !== undefined && (p.overall ?? 0) < filter.minOverall) return false;
    if (filter.maxOverall !== undefined && (p.overall ?? 0) > filter.maxOverall) return false;
    if (filter.minAge !== undefined && (p.age ?? 0) < filter.minAge) return false;
    if (filter.maxAge !== undefined && (p.age ?? 0) > filter.maxAge) return false;
    return true;
  });

  const sortBy = filter.sortBy ?? "overall";
  const dir = filter.sortDir ?? "desc";
  const factor = dir === "desc" ? -1 : 1;
  res.sort((a, b) => {
    if (sortBy === "name") return factor * a.name.localeCompare(b.name);
    const av = (a[sortBy as "overall" | "potential" | "age"] ?? 0) as number;
    const bv = (b[sortBy as "overall" | "potential" | "age"] ?? 0) as number;
    return factor * (av - bv);
  });

  if (filter.limit && filter.limit > 0) res = res.slice(0, filter.limit);
  return res;
}

export function playersByClub(players: Player[], club: string): Player[] {
  return findPlayers(players, { club, sortBy: "overall", sortDir: "desc" });
}

export function topBrazilianPlayers(players: Player[], limit = 10): Player[] {
  return findPlayers(players, {
    nationality: "Brazil",
    sortBy: "overall",
    sortDir: "desc",
    limit,
  });
}

File: experiments/adrianco/experiment-5/runs/language=go_model=claude-opus-4-8_tooling=none/rep2/internal/soccer/players.go
// players.go implements FIFA player-database queries.
package soccer

import (
	"sort"
	"strings"
)

// PlayerFilter describes search criteria for SearchPlayers. Empty fields are
// ignored. Name, Nationality, Club and Position use case- and accent-
// insensitive substring matching.
type PlayerFilter struct {
	Name        string
	Nationality string
	Club        string
	Position    string
	MinOverall  int
	SortBy      string // "overall" (default), "potential", "age", "name"
	Limit       int    // 0 = no limit
}

// SearchPlayers returns players matching the filter, sorted per SortBy
// (descending for ratings, ascending for name/age).
func (s *Store) SearchPlayers(f PlayerFilter) []Player {
	var out []Player
	for _, p := range s.Players {
		if f.MinOverall > 0 && p.Overall < f.MinOverall {
			continue
		}
		if !containsFold(p.Name, f.Name) {
			continue
		}
		if !containsFold(p.Nationality, f.Nationality) {
			continue
		}
		if !containsFold(p.Club, f.Club) {
			continue
		}
		if !containsFold(p.Position, f.Position) {
			continue
		}
		out = append(out, p)
	}
	sortPlayers(out, f.SortBy)
	if f.Limit > 0 && len(out) > f.Limit {
		out = out[:f.Limit]
	}
	return out
}

func sortPlayers(ps []Player, by string) {
	switch strings.ToLower(by) {
	case "name":
		sort.SliceStable(ps, func(i, j int) bool { return ps[i].Name < ps[j].Name })
	case "age":
		sort.SliceStable(ps, func(i, j int) bool { return ps[i].Age < ps[j].Age })
	case "potential":
		sort.SliceStable(ps, func(i, j int) bool { return ps[i].Potential > ps[j].Potential })
	default: // overall
		sort.SliceStable(ps, func(i, j int) bool { return ps[i].Overall > ps[j].Overall })
	}
}

// ClubSummary aggregates players grouped by club.
type ClubSummary struct {
	Club       string
	Count      int
	AvgOverall float64
}

// PlayersByClub groups the players matching the (optional) nationality filter
// by club, returning the summaries sorted by player count descending.
func (s *Store) PlayersByClub(nationality string) []ClubSummary {
	type acc struct {
		count int
		total int
	}
	groups := map[string]*acc{}
	for _, p := range s.Players {
		if p.Club == "" {
			continue
		}
		if !containsFold(p.Nationality, nationality) {
			continue
		}
		a := groups[p.Club]
		if a == nil {
			a = &acc{}
			groups[p.Club] = a
		}
		a.count++
		a.total += p.Overall
	}
	out := make([]ClubSummary, 0, len(groups))
	for club, a := range groups {
		avg := 0.0
		if a.count > 0 {
			avg = float64(a.total) / float64(a.count)
		}
		out = append(out, ClubSummary{Club: club, Count: a.count, AvgOverall: avg})
	}
	sort.SliceStable(out, func(i, j int) bool {
		if out[i].Count != out[j].Count {
			return out[i].Count > out[j].Count
		}
		return out[i].Club < out[j].Club
	})
	return out
}

// containsFold reports whether needle is an accent-insensitive, case-
// insensitive substring of haystack. An empty needle always matches.
func containsFold(haystack, needle string) bool {
	if needle == "" {
		return true
	}
	return strings.Contains(
		strings.ToLower(FoldAccents(haystack)),
		strings.ToLower(FoldAccents(needle)),
	)
}
